colossalai/nn/optimizer/distributed_adafactor.py

```diff
@@ -0,0 +1,308 @@
+    def _get_options(param_group, param_shape):
+        """
+        Determines whether the current param is factored
+        Args:
+            param_group : param group
+            param_shape : Original Shape of param
+
+        """
+        factored = len(param_shape) >= 2
+        use_first_moment = param_group["beta1"] is not None
+        return factored, use_first_moment
+
+    @staticmethod
+    def _rms(tensor):
+        return tensor.norm(2) / (tensor.numel() ** 0.5)
+
+    @staticmethod
+    def _approx_sq_grad(exp_avg_sq_row, exp_avg_sq_col):
+        r_factor = (exp_avg_sq_row / exp_avg_sq_row.mean(dim=-1, keepdim=True)).rsqrt_().unsqueeze(-1)
+        c_factor = exp_avg_sq_col.unsqueeze(-2).rsqrt()
+        return torch.mul(r_factor, c_factor)
+
+    # approx_sq_grad for row parallel weight
+    @staticmethod
+    def _approx_sq_grad_row_parallel(exp_avg_sq_row, exp_avg_sq_col, sq_row_meam):
+        # row_meam = sq_row_meam
+        r_factor = (exp_avg_sq_row / sq_row_meam).rsqrt_().unsqueeze(-1)
+        c_factor = exp_avg_sq_col.unsqueeze(-2).rsqrt()
+        return torch.mul(r_factor, c_factor)
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        """
+        Performs a single optimization steps
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+        """
+        param_groups: Dict
+        {
+            "params":[weight, bias]
+            "lr"
+            "eps"
+            "clip_threshold"
+            "decay_rate"
+            "beta1"
+            "weight_decay"
+            "scale_parameter"
+            "relative_step"
+            "warmup_init"
+        }
+        """
+        for group in self.param_groups:
+            # update weight & bias
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+                grad = p.grad
+                if grad.is_sparse:
+                    raise RuntimeError("Adafactor does not support sparse gradients.")
+
+                state = self.state[p]
+                self.grad_shape = grad.shape  # 1 dim shape
+                param_is_dtensor = is_distributed_tensor(self.shard_to_param.get(id(p)))
+                if param_is_dtensor:
+                    self.grad_shape = self.shard_to_param.get(id(p)).shape  # tp shape (2 dim)
+                self.factored, self.use_first_moment = self._get_options(group, self.grad_shape)
+
+                if len(state) == 0:
+                    state["step"] = 0
+                    if self.use_first_moment:
+                        # Exponential moving average of gradient values
+                        state["exp_avg"] = torch.zeros_like(p)
+                    if self.factored:
+                        self.shard_spec = get_sharding_spec(self.shard_to_param.get(id(p)))
```

Comment:
this also should be dict

Reply:
> this also should be dict

Fixed in https://github.com/hpcaitech/ColossalAI/pull/5484/commits/2ffca49595b4b6eb75c8260f7b6da63ce2758211;